src/app/rmmapi/rbwebmail.ts

```diff
@@ -207,40 +209,49 @@ export class RunboxWebmailAPI {
     }
 
     public deleteCachedMessageContents(messageId: number) {
-        if (this.messageContentsCache[messageId]) {
-            delete this.messageContentsCache[messageId];
+        this.messageCache.delete(messageId);
+        if (this.messageContentsRequestCache[messageId]) {
+            delete this.messageContentsRequestCache[messageId];
         }
     }
 
-    public getMessageContents(messageId: number, refresh = false): Observable<MessageContents> {
-        if (!refresh && this.messageContentsCache[messageId]) {
-            return this.messageContentsCache[messageId];
-        } else {
-            const messageContentsObservable = new AsyncSubject<MessageContents>();
-
-            this.messageContentsCache[messageId] = messageContentsObservable;
-
-            this.http.get('/rest/v1/email/' + messageId)
-            .pipe(
-                map((r: any) => r.result),
-            ).subscribe((r) => {
-                messageContentsObservable.next(r);
-                messageContentsObservable.complete();
-            }, err => {
-                delete this.messageContentsCache[messageId];
-                messageContentsObservable.error(err);
-            });
+    public getCachedMessageContents(messageId: number): Promise<MessageContents> {
+        let cached = this.messageContentsRequestCache[messageId];
+        if (!cached) {
+            // This kind of multilevel caching may seem excessive,
+            // but it turns out that pulling a single row from indexedDB (or maybe from Dexie specifically?)
+            // can easily take over 100ms: the difference between "instant" and "fast"
+            // Since instant is the only acceptable result, we cache indexedDB lookups here.
 
-            return messageContentsObservable;
+            cached = this.messageContentsRequestCache[messageId] = this.messageCache.get(messageId);
         }
+        return cached;
     }
 
```

Comment:
It doesn't seem to limit the messageContentsRequestCache in any way - it would probably be sane to have a max number of items in it (tracked by last used date?) .. just in case people scroll through pages of emails then complain that their browser explodes

Reply:
Add an LRU in 90625739e0c0fa45803f0358c9d5e91f087a6e93.
We could probably make it a little more generic for more varied uses if we decide that we need that sort of thing in other components. Currently it'll hold a maximum of 100 messages at a time (the rest will remain in IndexedDB), and enforce the cache limit no sooner than 1 second after the last insertion.